Modules/Sources/WordPressReader/Comments/Views/WebCommentContentRenderer.swift

```diff
@@ -0,0 +1,158 @@
+    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
+        // Wait until the HTML document finished loading.
+        // This also waits for all of resources within the HTML (images, video thumbnail images) to be fully loaded.
+        webView.evaluateJavaScript("document.readyState") { complete, _ in
+            guard complete != nil else {
+                return
+            }
+
+            // To capture the content height, the methods to use is either `document.body.scrollHeight` or `document.documentElement.scrollHeight`.
+            // `document.body` does not capture margins on <body> tag, so we'll use `document.documentElement` instead.
+            webView.evaluateJavaScript("document.documentElement.scrollHeight") { [weak self] height, _ in
+                guard let self, let height = height as? CGFloat else {
+                    return
+                }
+
+                /// The display setting's custom size is applied through the HTML's initial-scale property
+                /// in the meta tag. The `scrollHeight` value seems to return the height as if it's at 1.0 scale,
+                /// so we'll need to add the custom scale into account.
+                let actualHeight = round(height * self.displaySetting.size.scale)
+                self.delegate?.renderer(self, asyncRenderCompletedWithHeight: actualHeight)
+            }
+        }
+    }
+
+    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction) async -> WKNavigationActionPolicy {
+        switch navigationAction.navigationType {
+        case .other:
+            // allow local file requests.
+            return .allow
+        default:
+            guard let destinationURL = navigationAction.request.url else {
+                return .allow
+            }
+            self.delegate?.renderer(self, interactedWithURL: destinationURL)
+            return .cancel
+        }
+    }
+}
+
+private extension WebCommentContentRenderer {
+    /// Returns a formatted HTML string by loading the template for rich comment.
+    ///
+    /// The method will try to return cached content if possible, by detecting whether the content matches the previous content.
+    /// If it's different (e.g. due to edits), it will reprocess the HTML string.
+    ///
+    /// - Parameter content: The content value from the `Comment` object.
+    /// - Returns: Formatted HTML string to be displayed in the web view.
+    ///
+    func formattedHTMLString(for comment: String) -> String {
+        // remove empty HTML elements from the `content`, as the content often contains empty paragraph elements which adds unnecessary padding/margin.
+        // `rawContent` does not have this problem, but it's not used because `rawContent` gets rid of links (<a> tags) for mentions.
+        let comment = comment
+            .replacingOccurrences(of: Self.emptyElementRegexPattern, with: "", options: [.regularExpression])
```

Comment:
Ideally, I want to preprocess this in the background and store in Core Data in this way.